rasa/core/policies/intent_ted_policy.py

```diff
@@ -536,6 +552,117 @@ def _check_unlikely_intent(
 
         return False
 
+    @staticmethod
+    def _collect_label_id_grouped_scores(
+        output_scores: Dict[Text, Union[np.ndarray, Dict[Text, Any]]],
+        label_ids: np.ndarray,
+    ) -> Dict[int, List[List[float]]]:
+        """Collects similarities predicted for each label id.
+
+        For each `label_id`, we collect similarity scores across
+        all trackers and categorize them into two buckets:
+            1. Similarity scores when `label_id` is the correct label.
+            2. Similarity scores when `label_id` is the wrong label.
+
+        Args:
+            label_ids: Numerical IDs of labels for each data point used during training.
+            output_scores: Model's predictions for each data point.
+
+        Returns:
+            Negative similarities and minimum positive
+            similarity grouped by intent label id.
+        """
+        label_id_scores: Dict[int, List[List[float]]] = defaultdict(list)
+        unique_label_ids = np.unique(label_ids).tolist()
+        if LABEL_PAD_ID in unique_label_ids:
+            unique_label_ids.remove(LABEL_PAD_ID)
+
+        for label_id in unique_label_ids:
+            label_id_scores[label_id] = [[], []]
+
+        for index, all_pos_labels in enumerate(label_ids):
+
+            for candidate_label_id in unique_label_ids:
+                if candidate_label_id in all_pos_labels:
+                    label_id_scores[candidate_label_id][0].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+                else:
+                    label_id_scores[candidate_label_id][1].append(
+                        output_scores["similarities"][index, 0, candidate_label_id]
+                    )
+
+        # Get only unique scores so that duplicate
+        # trackers created because of permutation are pruned out.
+        unique_label_id_scores = {
+            label_id: [list(set(scores[0])), list(set(scores[1]))]
+            for label_id, scores in label_id_scores.items()
+        }
+        return unique_label_id_scores
+
+    @staticmethod
+    def _compute_label_tolerance_thresholds(
+        label_id_scores: Dict[int, List[List[float]]]
+    ) -> Dict[int, List[float]]:
+        """Construct
+
+        Args:
+            label_id_scores:
+
+        Returns:
+
+        """
+
+        def _get_threshold_for_tolerance(
+            negative_values, tolerance_percentile
+        ) -> float:
+            index = max(
+                0,
+                min(
+                    np.size(negative_values) - 1,
```

Comment:
So we don't allow tolerances > 1? Just checking.

Reply:
Nope, I've added a check in `train_utils.py` for it.